metabench.cmake

```diff
@@ -44,54 +50,68 @@ set(METABENCH_DIR ${CMAKE_CURRENT_BINARY_DIR}/_metabench)
-    target_include_directories(${target} PUBLIC "${template_dir}")
-    set_target_properties(
-        ${target} PROPERTIES
-        RUNTIME_OUTPUT_DIRECTORY "${target_dir}"
-        RULE_LAUNCH_COMPILE "${RUBY_EXECUTABLE} -- ${metabench_rb_path}"
+    # Add a dummy executable that will be used to collect the dataset.
+    # We'll build this executable multiple times for different values
+    # of `n`, and collect compilation statistics each time. Compiling
+    # through CMake allows us to be portable without additional work.
+    file(WRITE ${METABENCH_DIR}/${path_to}/${target}.cpp "")
+    add_executable(${target} EXCLUDE_FROM_ALL ${METABENCH_DIR}/${path_to}/${target}.cpp)
+    set_target_properties(${target} PROPERTIES
+        RULE_LAUNCH_COMPILE "${RUBY_EXECUTABLE} -- ${MEASURE_RB_PATH}"
+        RUNTIME_OUTPUT_DIRECTORY "${METABENCH_DIR}/${path_to}"
     )
+    target_include_directories(${target} PUBLIC "${CMAKE_CURRENT_SOURCE_DIR}/${path_to}")
 
-    # We pass the metabench.test.rb.in file through CMake's `configure_file`.
-    set(metabench_test_rb_path "${target_dir}/${target}.metabench.test.rb")
-    configure_file("${METABENCH_TEST_RB_IN_PATH}" "${metabench_test_rb_path}" @ONLY)
 
-    add_executable(${target}.test EXCLUDE_FROM_ALL "${source}")
-    target_include_directories(${target}.test PUBLIC "${template_dir}")
-    set_target_properties(
-        ${target}.test PROPERTIES
-        RUNTIME_OUTPUT_DIRECTORY "${target_dir}"
-        RULE_LAUNCH_COMPILE "${RUBY_EXECUTABLE} -- ${metabench_test_rb_path}"
+    # Add a command to generate the JSON file that will contain the measurements
+    # we collect for this dataset when we build the executable above.
+    set(configured_metabench_rb "${METABENCH_DIR}/${path_to}/${target}.metabench.rb")
+    configure_file(${METABENCH_RB_IN_PATH} ${configured_metabench_rb} @ONLY)
+    set(json_file "${METABENCH_DIR}/${path_to}/${target}.json")
+    add_custom_command(OUTPUT ${json_file}
+        COMMAND ${RUBY_EXECUTABLE} -r json -r ${configured_metabench_rb}
+            -e "range = (${range})"
+            -e "data = Metabench.compile_time('${CMAKE_CURRENT_SOURCE_DIR}/${path_to_template}', range)"
+            -e "IO.write('${json_file}', JSON.generate(data))"
+        DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/${path_to_template}
+        VERBATIM USES_TERMINAL
     )
 
-    add_test(
-        NAME ${target}
-        COMMAND ${CMAKE_COMMAND} --build . --target ${target}.test
-        WORKING_DIRECTORY "${PROJECT_BINARY_DIR}"
+
+    # We also setup a CTest target to test the generation of the dataset.
+    # We do not actually collect any data here.
+    add_test(NAME ${target}
+        COMMAND ${CMAKE_COMMAND} -E env METABENCH_TEST_ONLY=true
+            ${RUBY_EXECUTABLE} -r tilt/erb -r ${configured_metabench_rb}
+            -e "range = (${range})"
+            -e "data = Metabench.compile_time(\"${CMAKE_CURRENT_SOURCE_DIR}/${path_to_template}\", range)"
```

Comment:
wouldn't it be much easier to just pass `[range[0], range[-1]]` to Metabench.compile_time?